### PR TITLE
Fixes #666: Refresh Token cleanup

### LINKED ISSUE #666: Token Auth: Refresh token unexpected behavior
Please forgive the detailed nature of this bug. I'll include screenshots to try to make it more clear what is happening.

When an access token is generated (by visiting /login-token), the return contains both the `access` and `refresh` tokens.

screen shot 2015-10-02 at 12 30 47 pm

If I then use my refresh token to generate a new access token, the refresh token I use is deleted (as expected), but the old access token is not deleted (which is unexpected).

screen shot 2015-10-02 at 12 32 21 pm

When I then try to use my newly-created access token, I get unpredictable results (I get no refresh token and the old access token, even though I have a new access token):

screen shot 2015-10-02 at 12 32 46 pm

The expected functionality would be that after I've used a refresh token, when I then use the new access token, I would get back the paired tokens (access and refresh) and that the DB would be cleaned up appropriately.

**Proposed fix:** I think this can be easily accomplished by deleting the paired access token [when the refresh token is deleted](https://github.com/RESTful-Drupal/restful/blob/7.x-1.x/modules/restful_token_auth/plugins/restful/restful_token_auth/token_auth/refresh_token/1.0/RestfulRefreshTokenAuthentication.class.php#L50).

As an aside, it looks like there is no cleanup at all being done on the [entity reference table](https://github.com/RESTful-Drupal/restful/blob/7.x-1.x/modules/restful_token_auth/includes/RestfulTokenAuthController.php#L31).

### PR DESCRIPTION
**Pull request by lukewertz**
Delete the access token associated with a refresh token when a new refresh token is created. Also clean up the reference table.

### CODE REVIEW

File: modules/restful_token_auth/plugins/restful/restful_token_auth/token_auth/refresh_token/1.0/RestfulRefreshTokenAuthentication.class.php [hunk 1]
```diff
@@ -47,6 +47,22 @@ public function refreshToken($token) {
     // Remove the refresh token once used.
     $refresh_token = entity_load_single('restful_token_auth', key($results['restful_token_auth']));
     $uid = $refresh_token->uid;
+
+    // Get the access token linked to this refresh token then do some cleanup.
+    $accessTokenQuery = new EntityFieldQuery();
+    $accessTokenReference = $accessTokenQuery
```
Comment (e0ipso) [inline modules/restful_token_auth/plugins/restful/restful_token_auth/token_auth/refresh_token/1.0/RestfulRefreshTokenAuthentication.class.php]: Please change `$accessTokenReference` to `$access_token_reference` as per Drupal Coding Standards.

File: modules/restful_token_auth/plugins/restful/restful_token_auth/token_auth/refresh_token/1.0/RestfulRefreshTokenAuthentication.class.php [hunk 2]
```diff
@@ -47,6 +47,22 @@ public function refreshToken($token) {
     // Remove the refresh token once used.
     $refresh_token = entity_load_single('restful_token_auth', key($results['restful_token_auth']));
     $uid = $refresh_token->uid;
+
+    // Get the access token linked to this refresh token then do some cleanup.
+    $accessTokenQuery = new EntityFieldQuery();
```
Comment (e0ipso) [inline modules/restful_token_auth/plugins/restful/restful_token_auth/token_auth/refresh_token/1.0/RestfulRefreshTokenAuthentication.class.php]: Please change `$accessTokenQuery` to `$access_token_query` as per Drupal Coding Standards.

File: modules/restful_token_auth/plugins/restful/restful_token_auth/token_auth/refresh_token/1.0/RestfulRefreshTokenAuthentication.class.php [hunk 3]
```diff
@@ -47,6 +47,22 @@ public function refreshToken($token) {
...
+    $accessTokenReference = $accessTokenQuery
+      ->entityCondition('entity_type', 'restful_token_auth')
+      ->entityCondition('bundle', 'access_token')
+      ->fieldCondition('refresh_token_reference', 'target_id', $refresh_token->id)
+      ->range(0, 1)
+      ->execute();
+
+    if (!empty($accessTokenReference['restful_token_auth'])) {
+      $access_token = key($accessTokenReference['restful_token_auth']);
+      entity_delete('access_token', $refresh_token->id);
```
Comment (e0ipso) [inline modules/restful_token_auth/plugins/restful/restful_token_auth/token_auth/refresh_token/1.0/RestfulRefreshTokenAuthentication.class.php]: What is this for?

I believe that `'access_token'` is not an entity type, but a bundle.

File: modules/restful_token_auth/plugins/restful/restful_token_auth/token_auth/refresh_token/1.0/RestfulRefreshTokenAuthentication.class.php [hunk 4]
```diff
@@ -47,6 +47,22 @@ public function refreshToken($token) {
...
+    $access_token_reference = $access_token_query
+      ->entityCondition('entity_type', 'restful_token_auth')
+      ->entityCondition('bundle', 'access_token')
+      ->fieldCondition('refresh_token_reference', 'target_id', $refresh_token->id)
+      ->range(0, 1)
+      ->execute();
+
+    if (!empty($access_token_reference['restful_token_auth'])) {
+      $access_token = key($access_token_reference['restful_token_auth']);
+      entity_delete('access_token', $refresh_token->id);
```
Comment (e0ipso) [inline modules/restful_token_auth/plugins/restful/restful_token_auth/token_auth/refresh_token/1.0/RestfulRefreshTokenAuthentication.class.php]: What is this for?

I believe that 'access_token' is not an entity type, but a bundle.

Comment (lukewertz) [inline modules/restful_token_auth/plugins/restful/restful_token_auth/token_auth/refresh_token/1.0/RestfulRefreshTokenAuthentication.class.php]: You're right; my mistake. Thanks for catching it!

**Comment by lukewertz** [comment]
@e0ipso, I see that you tagged this with "needs 2.x port." I've not worked on the 2.x branch at all; would you expect me to write the 2.x code for this as well before this can get merged?

**Comment by josephdpurcell** [comment]
I'm interested in this PR as well. Any updates?

**Comment by e0ipso** [comment]
This is ready to be merged.

**Comment by e0ipso** [comment]
Merged. Thanks!!!

**Comment by e0ipso** [comment]
This was ported to 7.x-2.x in https://github.com/RESTful-Drupal/restful/pull/731. Removing tag.

**Pull request closed without merging**